designs/disruption-controls-by-action.md

```diff
@@ -0,0 +1,188 @@
+}
+
+type EmptinessSpec struct {
+    DisruptionCommonSpec
+}
+
+type Budget struct {
+    Nodes    string  `json:"nodes"`
+    Schedule *string `json:"schedule,omitempty"`
+    Duration *string `json:"duration,omitempty"`
+}
+}
+
+type DisruptionAction string 
+
+const (
+	All DisruptionAction = "All"
+	Consolidation DisruptionAction = "Consolidation" 
+	Drift DisruptionAction = "Drift" 
+	Expiration DisruptionAction = "Expiration" 
+	Emptiness DisruptionAction = "Emptiness"
+)
+```
+#### Example 
+
+```yaml 
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: example-nodepool
+spec:
+  disruption:
+    defaults:
+      budgets: 
+        - nodes: 10% 
+	  schedule: "0 0 1 * *"
+	  duration: 1h 
+    consolidation:
+      consolidationPolicy: WhenUnderutilized
+      disruptAfter: "30m"
+      budgets:
+	- nodes: "20%"
+	  schedule: "0 0 1 * *"
+	  duration: "1h"
+    drift:
+      disruptAfter: "1h"
+      budgets:
+	- nodes: "10%"
+	  schedule: "0 0 * * 0"
+	  duration: "2h"
+    expiration:
+      disruptAfter: "Never"
+```
+
+#### Considerations 
+Some of the API choices for a given action seem to follow a similar pattern. These include ConsolidateAfter, ExpireAfter, and there are discussions about introducing a global DisruptAfter. Moreover, when discussing disruption budgets, we talk about adding behavior for each action. It appears there is a need for disruption controls within the budgets for each action, not just overall.
+
+This approach aligns well with controls that apply to all existing actions. The proposal presented here is similar to the one mentioned above in relation to the actions we allow to be defined (All, Consolidation, Drift, Expiration, Emptiness).
+
+This proposal is currently scoped for disruptionBudgets by action. However, we should also consider incorporating other generic disruption controls into the PerActionControls, even if we do not implement them immediately. Moving ConsolidateAfter and ExpireAfter into the per-action controls is a significant migration that requires careful planning and its own dedicated design. This proposal simply demonstrates a potential model that highlights the benefits of defining controls at a per-action level of granularity.
+
+### Pros + Cons 
+* 👍 This model starts to make more sense as we continue to add general behaviors that apply to all disruption actions where users will want control on the action level.
+* 👍 Provides place per action for generic controls intended to be shared across all disruption methods. While not all methods share the same disruption actions, there are already cases for this. 
+* 👍 Could extend other fields beyond generic values for specific actions with validation   
+*  👎👎 Breaking API Change for Budgets at least, and if we decide to model DisruptAfter we also would have to break those apis. It might make sense to break the budgets now before they have garnered large adoption as it becomes harder to make the change as time goes on.
```

Comment:
I might be missing it, but I don't actually see what the required breaking change is here? What would be causing that?

I think my main reason for staying away from this approach is keeping the "default" budget case simple where folks just want to restrict disruption of any kind (or increase the number of nodes disrupted at a time), with the ability to make something more complex if they want.

Reply:
We would be moving the budgets api somewhere else so thats a breaking api change imo right? Also if we did this approach I would want to model ConsolidateAfter, and ExpireAfter into Generic Per Action DisrupAfter controls.